nvm.sh

```diff
@@ -3606,7 +3606,18 @@ EOF
 }
 
 nvm_supports_xz() {
-  if [ -z "${1-}" ] || ! command which xz >/dev/null 2>&1; then
+  if [ -z "${1-}" ]; then
+    return 1
+  fi
+
+  # macOS 10.9.0 and later support extracting xz with tar
+  if [ "$(nvm_get_os)" = "darwin" ]; then
+    MACOS_VERSION="$(sw_vers -productVersion)"
```

Comment:
should the previous line also check `nvm_has sw_vers`?
```suggestion
    local MACOS_VERSION
    MACOS_VERSION="$(sw_vers -productVersion)"
```
(the "local" line is needed for ksh)

Reply:
> should the previous line also check `nvm_has sw_vers`?

If you like, that's no problem. I have no strong opinion either way. 

The `sw_vers` utility has been in macOS/OS X since at least as long as there has been support for xz. [This page](http://sibr.com/info/sw_vers.html) indicates it goes back to OS X 10.3 with support for the `-productVersion` flag. Unless the user has deleted it, it is available, as a rule, in OS X/macOS that is at all recent. Also, if the program fails to run, and we can't confirm macOS is 10.9 or newer, the funtion exits 1 so we fall back to gzip. I don't think it adversely affects the branching logic in that case... IMO it's subjectively a bit over-engineered but not really a bad thing to check for it if you like.

If there's a particularly bad consequence to attempting to run `sw_vers` when it isn't on the PATH, or I suppose just to match the convention in rest of the script, I'd be open to changing it as you suggest.

Reply:
Actually, regarding if there is no `sw_vers` on the PATH: I'm not sure if we capture an error message about `_sw_vers` not existing, and try to compare the string of the error message as a version against 10.9.0 via `nvm_version_greater()`? That could be weird, I guess.

(macOS not having `sw_vers` would _also_ be weird, but desn't mean we can't plan for it I suppose.)

Not sure how `nvm_version_greater` handles letters in strings. I suppose I could test that real quick. Also not sure if the subshell `$()` would pass the error message if there is no `sw_vers` as a string to become the value of `MACOS_VERSION`. I can test that too.

Reply:
Semi-tangential point: it could be neat to test to some extent in a macOS Travis CI environment.

Reply:
If it's not too slow, we could certainly add a MacOS build to the test matrix.

Reply:
Alright, so I tested the (unlikely) situation if there is no `sw_vers` on the PATH and we are running on macOS. (FWIW, I tested this with the os check looking for `linux` instead of `darwin`, since I'm actually testing on linux...)

Assuming we don't gate the feature with `nvm_has sw_vers`, you get an error something like this:
```
./thing2.sh: line 64: sw_vers: command not found
```
Which I suppose is undesirable.

---

If there is no `sw_vers`, we can't detect whether there is support for xz, so should we use xz or gzip? gzip is conservative, but xz support is (I presume?) very common among current macOS users. (xz support goes all the way back to 10.9 Mavericks).

Alternatively, we could check for `liblzma.dylib` at (IIRC) `/usr/lib/liblzma.dylib`... so we don't have to worry about if `sw_vers` exists... realistically I would be shocked if a macOS system didn't have `sw_vers`...

Reply:
To summarize:
- It feels highly unlikely the user wouldn't have `sw_vers` on macOS
- If they don't, and we want to anticipate this, we should handle that scenario specifically.
  - fall back to gzip in this case?
- Alternative test would be to check for `/usr/lib/liblzma.dylib`, instead of checking the macOS version.
  - (or the actual path to that library if I am mis-remembering).
  - present means xz is supported, absent means xz is not supported.
  - IMO seems less elegant than the macOS version check, but that's subjective.

Reply:
Rather than doing version sniffing, is there a way we could *feature-detect* xz support on Macs, by piping a very tiny bit of xz-compressed data into tar and confirming if it decrypts properly?

Reply:
> _feature-detect_ xz support ...

Yeah, sure. I haven't done so before, but should be doable. How about compressing some `test.txt` with one line in it as some string, and check that it's the correct string after trying to decompress?

We can include `test.tar.xz` in the repo... (or make `test.txt` and compress/decompress it during the script?? Just to avoid shipping one more file?)

I would think to send all the command-line output from this process (if any) to /dev/null, since we just want to know if `tar` can decompress xz properly. The string comparison would be a way to really show it works, IMO.

Some compression wizard might have a cleverer suggestion, but the one above should work well enough. If we want to go a bit nuts, we could get a shasum of the decompressed `test.txt`, but I'm not familiar if xz preserves the exact original bits to that degree or not. One way to find out (prototype it out).

Reply:
Here's a prototyped-out xz feature test.

https://gist.github.com/DeeDeeG/46e1708c784f0fc207bf06624c409344

My main concern here is that obscure platforms ship their own `tar` that may or may not be able to handle the flags I use (`cJf` to create an archive, `xJf` to extract; `J` being a valid "use xz" flag for most implementations, but not sure if on all implementations. Should work on Linux and macOS though.)

Perf implication (disk usage): actually writes a text file, compresses to xz, decompresses again (writes output to another text file), and shasums both text files, then removes both the text files and the test xz tarball. Seems to happen just about instantaneously on my machine, though.

As written, it will decide the user can't extract xz (because we can't confirm yea or nay) if the user doesn't have all of `shasum`, `tar` and `cut` on the PATH (though other text manipulation utilities could be used instead of `cut`, if desired).

Reply:
I have a branch that implements the feature check: https://github.com/nvm-sh/nvm/compare/master...DeeDeeG:use_nvm_functions_xz

This breaks part of the [`nvh_supports_xz`](https://github.com/nvm-sh/nvm/blob/a1ad32e9cbc9ab2b1f955e0c7ab6e206cdfdd75c/test/fast/Unit%20tests/nvm_supports_xz) test, because that part of the test uses an empty "stub" file for the `xz` executable. We would want to let there be an actual `xz` executable available for the test to pass, given the changes I made.

Reply:
Here's a branch with a cleaned up implementation (feature detection, not OS detection), simple commit history, and passing tests.

https://github.com/nvm-sh/nvm/compare/master...DeeDeeG:xz-feature-test-rebased

Tests: https://travis-ci.com/DeeDeeG/nvm/builds/147113664 (Still waiting on the slower ones to finish, but the fast ones were the ones that were having problems, and those are passing already.)

If you prefer the direction that branch is going, I can force push that to this PR.

Reply:
I looked into making a version of this that uses `tar -J` and piping, without writing or reading any files to/from disk. This doesn't sem possible; Apparently `tar` can't create archives from input on stdin without referencing files on disk some way or other.

It's _is_ possible to only write a single text file to disk, and pipe that through `tar` so as to test xz compression/decompression, thus skipping writing the tar.xz archives to -- or reading them from -- the disk.

---

Here's an updated version that uses piping, other than having to write a small text file out to disk, so as to be able to pipe said text file through `tar`: https://github.com/nvm-sh/nvm/compare/master...DeeDeeG:xz-piping

Reply:
My thinking was to hardcode both a compressed and uncompressed string inside nvm.sh, and not to use separate files. Why does tar require a file to live on the disk?

Reply:
> Why does tar require a file to live on the disk?

See: https://askubuntu.com/a/587046

> `tar` _cannot_ read in file data from stdin for creating an archive, since then `tar` will have no way of knowing what a file is - where one begins or ends, what its path and metadata is.

Apparently just a stream of raw bits/other data does not contain enough information to create files/an archive from, since (I suppose) `tar` theoretically might be receiving multiple files, or just ASCII characters, or binary bits... etc. or some combination thereof -- whereas its task is to store these logically and reliably as the separate files they are intended to be. `tar`  (or `tar`'s coders) considers just a pipe of data too ambiguous to make an archive out of.

Reply:
Well there is a workaround... We could pipe in a complete file from some other command that sends an entire file, rather than just echoing a string. For example, I believe you can do `curl [URL] | tar`. But using `curl` implies getting a remote file from somewhere. So, not sure if that helps at all.

It's unclear to me if you can generate an entire file, metadata and all (filename, modification date, etc.) in the shell, without it ever hitting the disk, only being moved around through pipes.

Reply:
I was thinking of extracting. You can indeed _extract_ an archive fed in via pipe, but not _create_ an archive with files piped in.

Reply:
Short of patching `tar`, the feature to create archives based on piped input simply isn't available in `tar`.

Reply:
I can understand that writing a tiny file to the user's disk every run isn't desirable.

The two apparent alternatives are, as far as I can tell:
- Check support (with genuine feature detection) in some way other than creating an archive on the fly with `tar`.
  - Ship a basic archive to extract (included when the user downloads/installs `nvm`)
  - Download a basic archive off of the network, with e.g. `curl`
  - Cut `tar` out of the loop; Guess what extractor utility `tar` will use and just test that that works (`xz` binary on Linux, `gzip`/`gunzip` on macOS coincidentally should work.)
    - (Despite the name, algorithms other than gzip were rolled into `gzip`/`gunzip` on macOS, for whatever reason.)
- Use heuristics such as OS version and/or presence of an `xz` binary to "guess" xz support.
  - This is what `n` and `nvh` did in the PR I worked on, and it seems to work.
  - `xz` support is usually shipped with the OS these days (especially macOS is consistent with this), and the `xz` binary check is easy enough on Linux where things are a bit more unpredictable.
    - Given that xz support is shipped with all new macs, and baked into `tar` itself, it is pretty reliably there. Checking for macOS 10.9 or newer is nice, to keep `nvm` working on rather old, unsupported macOS releases (those released in 2012 or earlier). But even that is almost overkill in a way. Most users are on newer macOS than OS X Mountain Lion.
    - Everybody seems to use GNU tar when on Linux. New enough GNU tar (1.22+) requires an `xz` binary on the PATH to use the `-J` flag, or a command using the `-J` flag simply fails saying `xz` is missing. We do a greater level of checking than GNU `tar` does if we check for an `xz` binary on the PATH  on systems where GNU `tar` is used.
      - I guess somebody could have an ancient (pre-2009) GNU tar AND have an `xz`
 binary on the PATH... A very rare scenario. The VAST majority of people have newer `tar` _and_ `xz` on the PATH. Those on old out-of-support distros with ancient `tar` hopefully don't have `xz` binaries on the PATH.
        - Tangentially: I tested my `xz-piping` branch on Debian Lenny, and `nvm_platform_supports_xz()` from that branch actually does disable xz usage on really old `tar`. Reason being, the `tar` extract step errors out (`-J` isn't a  valid flag in really old `tar`), and puts nothing on stdout. So `EXTRACTED_STRING` becomes empty, whereas the string to compare it to is non-empty: "The quick brown fox jumps over the lazy dog." When these fail to match, the overall feature check fails, and xz usage is gated off by `nvm_platform_supports_xz()`.

(Sorry for the many comments. Also sorry this one is so huge, but I had a lot of thoughts to get out. Trying to cover all the bases and be thorough.)

Reply:
I'm not sure how to proceed; I prefer a robust feature check, but it'd be ideal to minimize the file operations nvm performs.

Reply:
AHA!

I figured something out.

We can inline the tarball (as hexadecimal raw data) into the script.

Here's a branch (showing comparison to `master`), let me know if you want it pushed here: https://github.com/nvm-sh/nvm/compare/master...DeeDeeG:xz-binary-piping

Technical explanation below (mostly just putting this here because I'm excited I figured it out, feel free to skip):

---

I encoded an xz-compressed tarball (the smallest size-on-disk of such a tarball I could make) into hex (performed a hex dump of the tarball) on my local machine by doing the following:

```sh
echo "a" > a
tar cJ a | xxd -p
```

(`xxd` is [apparently](https://stackoverflow.com/a/6292701) the command-line hex dump tool of choice for UNIX-likes?)

And then I just properly formatted the lines of the hex dump into a variable assignment in the script (setting the variable `HEX` to a string, which consists of the lines of the hex dump).

From there, the variable/string can be `echo`ed into the `xxd` hex dump utility in reverse mode. This give us binary being piped around that represents the xz-compressed tarball. This can be piped into `tar` for extraction, and the output should be ASCII `a`.

We can then string compare this output to the hard-coded basic string `a`, in a simple `if [ "${EXTRACTED_STRING}" = "a" ]`

And there we have the feature detection performed, without shipping any additional files, or reading from/writing to separate files on disk.

---

**_By the way_**, I get the following when manually running the `nvm_supports_xz` unit test:

```
$ ./nvm_supports_xz 
./nvm_supports_xz: 3620: xxd: not found
./nvm_supports_xz: 3620: tar: not found
```
This can be fixed, though it's just a tidying up thing for local testing. (These errors don't show up on Travis CI, just locally. And the test still passes, both locally and in CI.)

Reply:
That looks like exactly what I was hoping for. I'm not sure about `xxd` tho, is that in POSIX?

Reply:
`xxd` is part of `vim`, so the internet says. Not in-and-of-itself a part of POSIX.

I am fairly stumped on how to do this without `xxd`, because its "reverse mode" is very handy for generating binary data to pipe around.

I can pipe binary data around with `printf`, but I'm finding it hard to actually convert all of the hexdump (all the hexadecimal characters) to binary without outputting them to a file, or accidentally converting the binary to ASCII string representations of the binary or the hex.

I can format two characters of hex into binary at a time with `printf \xHH` where each `H` is a hex digit. But storing this anywhere but in a file seems to convert it to full-fat ASCII data, not raw binary data.

There are some nifty alternatives like `base64` which also isn't POSIX... And there seems to be a proper way to do it that is POSIX (`uuencode` and `uudecode`), but neither of these are installed by default in Ubuntu. (`uuencode` and `uudecode` are available in the Ubuntu repos as part of the `sharutils` package.)

Reply:
HMMMMM...

This is IMO extremely inelegant, but it works:

```sh
printf "\xfd\x37\x7a\x58\x5a\x00\x00\x00\xff\x12\xd9\x41\x02\x00\x21\
\x01\x1c\x00\x00\x00\x10\xcf\x58\xcc\xe0\x27\xff\x00\x5d\x5d\x00\x30\
\x80\x33\x50\x18\x68\x89\x30\x16\xd4\xad\x2b\xdb\x35\x32\xd1\xda\x84\
\x56\xc1\x98\x00\xca\x7c\x3f\xa0\xf7\x94\xed\xff\xcd\xc0\x02\xea\x3e\
\x71\x5b\xe7\x37\xdc\xd0\xe6\x0f\x47\x55\xe5\xdf\xcc\xf5\x59\xd1\xce\
\xbf\x59\xe8\xb9\x88\xa8\x68\x02\x9b\x8c\x55\x8a\x11\xbc\xb1\x3e\x4e\
\xc1\xf5\x98\x85\x0c\x8a\x37\xb7\xc0\xd1\x16\xfa\xc4\x71\x9d\x5a\x11\
\x3a\x7c\x89\x04\x4a\x00\x00\x00\x00\x00\x00\x00\x01\x71\x80\x50\x00\
\x00\x00\x89\xb5\xc0\x79\xa8\x00\x0a\xfc\x02\x00\x00\x00\x00\x00\x59\
\x5a" | tar xJ -O
```
(Outputs "a")

I mean it's elegant, _in a way..._ just a single command piped to another, but on the other hand... the escape characters take up _exactly as much space_ on disk as _the data itself_. HMMMMM.

Clever/elegant or not, in any case, this shows "It can be done" with just POSIX.

Reply:
let's do that then!

Reply:
This is giving me problems in `dash` for some reason.

(I understand `printf` to be a shell builtin, so perhaps the implementation is different.)

The problem is: When I try to run this from `dash`, I get ASCII output like so:

command: `printf "\xfd"`
output: `\xfd`

(Output should be this, in raw binary: `11111101` which shows as a "question mark in a diamond" placeholder glyph in my terminal.)

Reply:
I can reproduce this; I'm not sure how to address it yet nor why it happens.

Reply:
I wonder if it rises to the level of a bug for `dash`, or if they intentionally don't have a very advanced `printf` so as to KISS, and to marginally improve performance, like so many things in dash are.

Reply:
Some have suggested it's because that's not a valid utf-8 character.

Reply:
I think the "question mark in a diamond" goes with invalid or wrongly-encoded characters. And I suppose binary data would be indistinguishable from garbled or wrongly-encoded text.

But in shells other than `dash` are okay with my `printf` command... the output from my `printf` command _should be_ valid as binary data, but non-valid as text. So the question mark symbol is expected if all is working properly, as far as I am concerned.

In other words, seeing the question mark in the diamond is more on the right track than seeing plain text. Seeing plain text means we are feeding text into `tar` rather than proper binary data. This situation causes `tar` to fail in `dash`, as it (rightly) complains it can't extract the data we are giving it.

Reply:
(By the way, here are some Travis tests showing this working everywhere but `dash` and `sh`... On recent Ubuntu, `sh` is a symlink to `dash`, so they are the same shell in Ubuntu/the default Linux Travis CI environment. https://travis-ci.com/DeeDeeG/nvm/builds/147819367)

Reply:
I suppose we could branch the code; if the user's shell is `dash`, try something else (like `xxd`?)

Reply:
Another alternative might be to use `awk`?

Reply:
Maybe so. I don't know how to use it, though. I tried learning it a bit today, but I didn't get very far yet.

If someone else wants to provide an implementation, that would certainly move this PR forward. And I may try to continue learning `awk` over time and see where I get with it.

Reply:
As far as things I now how to do at the moment:
- Heuristic/educated guess approach (not a feature test)
- Feature test (not strictly POSIX)
- Leave as-is

Reply:
A non-POSIX feature test that does not have any false failures - ie, any place that doesn't support the feature test also, actually, does not support the feature being tested - then that should be fine. That's the important thing - that a feature test for X returns the correct boolean based on actual support for X.

Reply:
The most foolproof test I came up with so far was the one that required a small text file or a small xz-compressed tarball to be read from disk. (And if it isn't shipped with `nvm`, that implies we write it to and delete it from disk during the script as well).

In terms of a non-strictly POSIX test, the one that requires `xxd` works fabulously, no reads/writes to/from disk required, but I would have to do research that confirms how widely that utility is available. Particularly, I would like to see that `xxd` has been shipped by default with macOS for a long time. (More non-POSIX alternatives appear to include perl and python).

IMO some false negatives are a given. Supporting everybody flawlessly is hard or impossible to achieve, especially while also eliminating any implementation drawbacks. Falling back on gzip is not so bad a scenario, for the few who we can't determine their support, IMO.

(Indeed, the existing implementation has existed without complaint for some time. My only aim is for an incremental improvement, not perfection.)

Does anyone write in to say that they are trying to use `nvm` on BSD? Or any other non-Linux UNIX-like of some kind? If users are strictly on macOS or Linux, it is easier to reason about who needs what from the test behavior. (But universal-ish support can still be a design philosophy, so I defer to your judgment.)

(Re: other platforms: I see that Windows isn't supported, per the README, and is nonetheless covered to some extent by Linux support, by way of WSL. SunOS and AIX seem to be extremely niche, and `nvm` appears to be untested on these OSes).

Reply:
OK - if everywhere that this particular feature test (using `xxd`) should return `true`, also has `xxd`, i'm fine with saying that this test returns `false` when `command xxd` does not succeed.

Reply:
I'm awaiting a good opportunity to borrow a friend's mac computer, which has VM's with fresh OS installs going back to OS X 10.6 "Snow Leopard." This was the machine I did macOS research on for `nvh`/`n`.

I can test a variety of Linux distributions (including some rather old ones) on my current machine.

Reply:
HMM.

- macOS all the way back to 10.6 has `xxd`, so the test works on macOS.
- Most Linux distriutions ship with `xxd`, however:
  - Fedora and Manjaro both do not whip with `xxd` out of the box.

Reply:
If `xz` is present, then it seems like we wouldn't need this fallback `tar` check at all, no?

Reply:
On Linux, that is correct. GNU tar (new enough to be aware of the xz format) just uses `xz` from the PATH.

---

macOS is different:

macOS uses `bsdtar` from libarchive for its implementation of `tar`. libarchive `bsdtar` relies on something other than the `xz` binary; fairly certain it uses /usr/lib/liblzma.dylib

_(Side note: lzma and xz are slightly related technologies, so support for both can reasonably be provided by one library. I think it's even from [the same project](https://tukaani.org/xz/) that makes the `xz` binary.)._

**macOS' `tar` implementation (`bsdtar` from libarchive) _will not_ in any circumstance use the `xz` binary on the path, at all. Only `liblzma`.**

Some macOS users _do_ have `xz` on the PATH, and I have heard anecdotally that many of them got it via homebrew.

**macOS 10.8 or older with `xz` on the PATH would come up as a false-positive** for "supports xz [via `tar`'s extract mode]". It actually would not support it, because the `tar` shipped with older macOS is not xz-aware, and relatedly, the older OS does not ship `liblzma` to actually extract tarballs with.

---

My personal recommendations, considering just macOS and Linux:
- On Linux, checking for `xz` on the PATH is quite adequate, in my experience.
- On macOS, checking for macOS 10.9 or greater is quite adequate in my experience.
  - You could alternatively check for `/usr/lib/liblzma.dylib`, which might or might not be faster than running `sw_vers -productVersion`, I haven't benchmarked it. But should pass/fail the same OSes.)

Exceptions to these checks (false positives) would be someone manually replacing the `tar` shipped with their OS to some version that doesn't support xz. (Or obtaining an `xz` binary on a really old Linux distro (from 2009 or earlier) whose `tar` is too old to support `xz`.) I don't know if it's fair for anyone to support that scenario, and it feels unlikely to come up.

As I mentioned somewhere, Linux distros that have `xz` in the repos generally ship `tar` new enough to use it.

---

I personally root for such a "heuristic" approach as I described in my bullet points above, since it's an extremely light check, performance-wise, and quite accurate enough, practically speaking, in my experience.

---

P.S.: a more thorough (but IMO overkill) check could be:

- determine `tar` is GNU tar
  - determine GNU tar version is 1.22 or newer (version where xz support was added)
    - check for `xz` binary on PATH
OR
- determine `tar` is `bsdtar` from libarchive
  - determine libarchive is 2.7 or newer (version xz support was added)
    - if macOS, check for /usr/lib/liblzma.dylib OR
    - if macOS, determine macOS is 10.9 or greater
      - (macos 10.7 and 10.8 ship a new enough version of libarchive `bsdtar` that could have supported xz, but Apple configured it with the xz support disabled at compile time!)
      - (I have no idea how libarchive `bsdtar` behaves outside of macOS, or if it is even used on, say, actual straight-up BSD.)

Reply:
in very brief: the `xz` on PATH check doesn't work on macOS.

Reply:
hmm - if `tar` invokes `xz` on the PATH in implementations where this is reliable (ie, not mac) then could we test this by doing something like `[ "$(xz() { return 42 } ; tar --force-xz ; echo $?)" = '42' ]`?

Reply:
I think that gets hung up on some stuff I've run into while prototyping so far.

Mostly that you can't actually run`tar` (and get anything but an error), without creating or extracting a real tarball.

---

I take the brackets in that snippet `[]` to mean `test`, with running some stuff and echoing out its exit status on the left... then string comparing the exit status on the left with the string "42" on the right.

Such a check would hypothetically pass if the left side exited with status `42`.

Looking closer at the left side: I understand the `xz()` part as a function that is defined to just exit with status `42`. And then we run `tar xJ` or something like that to force `tar` to use xz compression.

we can't run `tar` without actually feeding it a tarball, or having it actually create a tarball (perhaps to stdout), or it will exit with an error. I didn't see anywhere in your snippet that creates or extracts a real tarball, so I think the exit status of `tar` will just be 1 or 2, not 42.

The left side would probably exit 1 or 2, and the string comparison would fail/the check would fail.

Reply:
ah, gotcha

Reply:
if we want to feature test `tar`, in an OS-agnostic/multi-OS way, and have that pass on `dash` (which btw is not shipped with macOS as far as I can tell) then I believe we need to make a small file on disk. Perhaps under NVM_DIR. Then we can compress/uncompress that with `tar`.

(This bypasses the above problem with `xxd` not being shipped with certain Linux distros).

If not doing a feature test, I think the most practical check is: macOS 10.9 or newer (pass), or some other os and `xz` on the PATH (pass). else (fail).

Reply:
if we do the feature check strictly on macOS, we can assume `xxd` is available and do that whole thing with piping some hex into `xxd`. For that matter, I think `dash` isn't really a thing on macOS. its name literally means the Debian Almquist Shell. It's a variant of `ash`... but again I don't think `ash` or `dash` are really used on macOS. So we could just use `printf` to print the hex as binary, which means we don't need `xxd` and we're POSIX-compliant again.

and not feature check on other OSes, just look for `xz` on PATH. which IMO is good enough, and it's what master branch already is doing.

Reply:
By the way, the "holy grail" IMO, in terns of implementing what you've said you hope for, would be some POSIX-compliant way to output binary data (in our case a tarball) to a pipe so `tar` can extract it.

`xxd` was close, but not available on some Linux distros. And `printf` works most of the time, but in a strictly bare-bones POSIX sense (as `dash` shows) it's not able to print out binary data.

If you find out about a way to get binary data into a pipe in the script, that works in all shells on all OSes we care about (ideally within POSIX), I will happily use it. I have tried to find such a thing, but so far, no luck.

Reply:
There are supposedly perl or python scripts that will do it, though. Maybe everywhere we care about has perl or python?

Reply:
I have dash installed on my mac (for testing), fwiw.

Reply:
We definitely can't rely on perl and/or python existing :-/

Reply:
IMO, as a pragmatist: the "`xz` on the PATH?" check has been good enough so far. There is inherently a safe fallback, which is gzip.

Furthermore, the environments people who download node in pretty uniformly support `xz` these days.

For comparison to a real-world parallel: `n` knows to check for macOS 10.9 or greater, rather than `xz` on the PATH, for macOS, as of \*checks notes\*... 16 days ago. And no-one has complained about it so far. In fact... no-one complained when `n` had no support for xz compression at all.

The only actually critical problem an xz check could have would be to return a false positive. Falling back to gzip is basically always okay.

I believe a simple check is good enough, and would be accurate. And one does not have to support exotic, non-default scenarios like a user overwriting the default system `tar`. So my view, I suppose this will be the last time I plainly reiterate it, but my view is the "`xz` on path?" check is good enough for Linux and SmartOS, a "macOS 10.9 or greater?" check is good enough for macOS. (Literally no idea about AIX, but this wouldn't _hurt_ AIX either, so it's arguably a moot point).

So that's what I'd suggest. Otherwise this PR will be stalled indefinitely, for want of a POSIX-compliant feature check, with no reads/writes to/from disk, that works in popular Linux distros _and_ macOS, and perhaps on SmartOS and AIX. But If we're waiting on that I'd like to officially pause my involvement with this PR until there is news on that front.

Thanks for working on this with me, but I am out of new ideas for now. Best of luck with the project, and best regards, in general. :+1:

Reply:
To be clear, as long as the outcome is always "the user can uncompress the archive and install node", i'm not particularly concerned with "the archive is the smallest possible one available" - i'm just trying not to regress performance :-)

Reply:
Some quick performance numbers; `nvm_supports_xz()` takes this long on the following branches on my main machine:

- `master`: 5 milliseconds
- `adjust-xz-platform-checks`: 8 or 9 milliseconds
  -  (with macOS version check, **_but_** version check code path is skipped, due to testing on Linux)
- `c`: 14 milliseconds
  -  (with piping hex into `xxd` then into `tar`)
- `tmp-file-tarball`: 21 to 23 milliseconds
  -  (with creating a file "a", containing the character "a"; creating and extracting an xz-compressed tarball from the file "a" with piping, then deleting "a")

---

I got these numbers by copy-pasting the functions from `nvm.sh` needed to run `nvm_supports_xz()`, and `nvm_supports_xz()` itself, once for each branch. (Although only `nvm_supports_xz()` and `nvm_platform_supports_xz()` were different among the branches, so I just copied and edited the first extracted mini script for the other two branches...) And then running `time ./[mini-script-filename].sh`

Impact should be roughly **double** this, though, as `nvm_supports_xz()` is typically run twice, back-to-back, in the script. Once when determining what compression type to use when extracting the downloaded tarball (stored as `tar_compression_flag` for later use) in `nvm_install_binary()` or `nvm_install_source()`, and once again when deciding what compression type of tarball to download in the first place in `nvm_download_artifact()`.

For OSes/platforms combos where a binary is not available, e.g. 32-bit macOS, `nvm_supports_xz()` is run up to **four times**. Once in `nvm_install_binary()` and consequently `nvm_download_artifact()` -- the download URL for 32-bit macOS 404's since there is no binary for 32-bit macOS for most versions of node these days. So we get "Binary download failed, trying source."... and then run `nvm_supports_xz` twice again, once in `nvm_install_source()` and once more in `nvm_download_artifact()`.

Reply:
It would be totally fine to cache the result of `nvm_supports_xz` for the life of a single `nvm` command (just not beyond that). That seems like a useful performance improvement independent of this PR.

Reply:
I tried the most perf-sensitive test I could think of for macOS.

Note: results are across the board slower on this machine, it's not the script, it's the slower computer I have for running macOS.

Here's a comparison of the run time for the `nvm_supports)xz()` function (and supporting functions) from `master`, versus a version that also checks for "OS is Darwin," and if so, "`/usr/lib/liblzma.dylib` is a file" (`[ -f /usr/lib/liblzma.dylib ]`):

- `master`: 22 to 26 milliseconds typically (one 28 millisecond outlier run)
  - with the "_system has xz on PATH_" code path active, because I put a dummy `xz` file at `/usr/local/bin/xz` with execute permissions 
- `basic_macOS_check`: 27 to 31 milliseconds typically (one 42 millisecond outlier run)
  - with the "_system is macOS and has liblzma.dylib_" code path active, because I'm testing on macOS with `/usr/lib/liblzma.dylib` present

Tried to make it apples to apples, in case all those version comparisons later in the script might slow it down. Didn't want to skip them; The target scenario is if the checks pass. If the tests fail, then `nvm_supports_xz()` should theoretically return 1 even quicker than it would return 0 after passing the checks throughout the whole function.

---

For fun, rather than running `nvm_supports_xz 10` at the bottom of my mini-script, I tried running through all the version check logic, which happens for any version greater than `0.12.10` but less than `4.0.0`. So I ran `nvm_supports_xz 2.3.3` instead.

`master`: 52 to 54 milliseconds (outliers at 65 and 119 milliseconds)
`basic_macOS_check`: 55 to 61 milliseconds

Reply:
code snippet used for `basic_macOS_check` speed tests above:

```sh
nvm_supports_xz() {
  if [ -z "${1-}" ]; then
    return 1
  fi

  if [ "$(command uname -s)" = 'Darwin' ] && ! [ -f "/usr/lib/liblzma.dylib" ]; then
    return 1
  elif ! command which xz >/dev/null 2>&1; then
    return 1
  fi

# [ . . . rest of nvm_supports_xz() as it appears in master . . . ]
```

Reply:
A gist of the actual scripts I ran, if you want to try to reproduce.

https://gist.github.com/DeeDeeG/2f32f818991cef953958446cd7aa70f1

(I ran `time ./[script_name.sh]` and took the "real" value, since waiting for disk or what-have-you is still relevant to the user experience. "Userland" CPU or kernel CPU time basically aren't relevant by themselves, in a vacuum, IMO.)

Reply:
Sorry for all the comments at once... Realized my check can still false positive on older macOS where we don't have `liblzma.dylib`, IF the user also has `xz` on the PATH such as from homebrew. Updated so the `xz` on PATH check only happens if not on macOS.

Had to redo the performance numbers, so:

For `nvm_supports_xz 2.3.3`:

- `master`: 51 to 52 milliseconds (one outlier run was 53 milliseconds)
- `basic_macOS_check_update`: 52 to 54 milliseconds

For `nvm_supports_xz 10`:

- `master`: 22 to 24 milliseconds (one outlier was 29 milliseconds)
- `basic_macOS_check_update`: 24 to 25 milliseconds (one outlier was 26, another was 37 milliseconds)

---

For fun, `master` with no `xz` on the PATH, which results in nvm just falling back on gzip:

- `nvm_supports_xz 10`: 16 to 17 milliseconds (one outlier was 20 milliseconds)

Reply:
New snippet used for `basic_macOS_check_update` above (this time with no apparent false positive scenarios):

```sh
nvm_supports_xz() {
  if [ -z "${1-}" ]; then
    return 1
  fi

  if [ "$(command uname -s)" = 'Darwin' ]; then
    if ! [ -f "/usr/lib/liblzma.dylib" ]; then
      return 1
    fi
  elif ! command which xz >/dev/null 2>&1; then
    return 1
  fi

# [ . . . rest of nvm_supports_xz() as it appears in master . . . ]
```

(Also added to the previously linked [gist](https://gist.github.com/DeeDeeG/2f32f818991cef953958446cd7aa70f1), which again is [here](https://gist.github.com/DeeDeeG/2f32f818991cef953958446cd7aa70f1).)